main.go

```diff
@@ -19,105 +18,119 @@ type FileInfo map[string]time.Time
 var watchedFiles = FileInfo{}
 
 func main() {
-	defaultPath, _ := os.Getwd()
+	defaultPath, err := os.Getwd()
+	if err != nil {
+		log.Fatal("could not get current directory::", err)
+	}
+
 	rootPath := flag.String("path", defaultPath, "the root path to be watched")
-	interval := flag.Int("interval", 5, "the interval (in seconds) for scanning files")
+	interval := flag.Duration("interval", 5*time.Second, "the interval (in seconds) for scanning files")
 	flag.Parse()
 
-	watchedFiles = walk(rootPath)
-	scheduleScanAt(rootPath, interval)
-}
+	if *interval < 0 {
+		log.Fatal("invalid interval, must be > 0:", *interval)
+	}
 
-func scheduleScanAt(rootPath *string, interval *int) {
-	runTicker := time.NewTicker(
-		time.Duration(time.Duration(*interval) * time.Second))
-	for {
-		select {
-		case <-runTicker.C:
-			scan(rootPath)
-		}
+	if err := os.Chdir(*rootPath); err != nil {
+		log.Fatal("could not change directory:", err)
+	}
+
+	watchedFiles = walk(rootPath)
+	for range time.NewTicker(*interval).C {
+		scan(rootPath)
 	}
 }
 
 func scan(rootPath *string) {
+	modifiedDirs := make(map[string]bool, 0)
 	for filePath, mostRecentModTime := range walk(rootPath) {
-		if lastModTime, ok := watchedFiles[filePath]; ok {
-			if lastModTime != mostRecentModTime {
-				watchedFiles[filePath] = mostRecentModTime
-				runOrSkipTest(filePath)
-			}
-		} else {
+		lastModTime, ok := watchedFiles[filePath]
+		if !ok {
 			// files recently added
-			fileInfo, _ := os.Stat(filePath)
+			fileInfo, err := os.Stat(filePath)
+			if err != nil {
+				log.Print("Stat:", filePath, err)
+				continue
+			}
 			watchedFiles[filePath] = fileInfo.ModTime()
+			continue
+		}
+
+		// no changes
+		if lastModTime == mostRecentModTime {
+			continue
+		}
+
+		watchedFiles[filePath] = mostRecentModTime
+		if !needsTest(filePath) {
+			continue
 		}
+
+		pkgDir := path.Dir(filePath)
+		modifiedDirs[pkgDir] = true
+	}
+
+	if len(modifiedDirs) == 0 {
+		return
 	}
+
+	dedup := make([]string, 0)
```

Comment:
By changing the data structure of `modifiedDirs` from `map` to `slice`, I think this `dedup` logic can be removed. You can call `test(modifiedDirs)` directly.